### ISSUE TITLE
onRowSelected called every time table is rendered

### ISSUE THREAD

**Issue by mercurjl** [jbetancur/react-data-table-component#320]
## Describe the bug
Every time my data table's data is changed the table is calling the function handling onRowSelected.  I have my data table in a class function mapped like so:
```
<DataTable
   columns={columns}
   data={this.state.filteredFlattenedReports}
   selectableRows
   onRowClicked={e => this.showModal(e)}
   defaultSortField={"date_created"}
   defaultSortAsc={false}
   onRowSelected={this.handleRowSelected}
   pagination={true}
   noHeader={true}
   selectableRowsPreSelectedField="selected"
/>
```
The handler, `this.handleRowSelected`, is what is called every time the table renders/data is changed. This causing my app to run `this.handleRowSelected` when it doesnt need to.

## To Reproduce
1. Create a data table with a data source and an onRowSelected handler
2. Notice that when the the data fed to the table changes, onRowSelected is called

## Expected behavior
onRowSelected should only ever fire when the user actually selects and item on the table

## Versions (please complete the following information)
 - React
 - Class Component
 - Windows
 - Firefox
 - 3.10.2

## Additional context
If it helps at all what I'm trying to accomplish is filtering based on user text input. There is a text input for every column, and as the user types it will filter out the data feeding the data table. This is how I noticed this in the first place, as I was typing in a filter and changing the table data my console logs in the onRowSelected handler would fire.

The underlying issue I'm attempting to solve in my app is maintaining what is selected in the data table while filters are being applied and removed.

**Comment by jbetancur** [comment]
Thank you for the detailed issue! Very helpful.

I've confirmed that this is a bug when using selectableRows and when the data changes (in your case filtering).

I'll have a fix soon...

**Comment by jbetancur** [comment]
I’m going to cut a release To fix the errant firing but there is still a related bug where if your data changes then all your selected rows get unselected. Honestly I’m not sure the correct way to handle when using selectable rows and the underlying data change.

In the meantime I’ll cut a 3.10.3 release soon as I can.

**Comment by mercurjl** [comment]
Awesome sounds good! Thanks for the quick action with this

**Comment by jbetancur** [comment]
3.10.3 is now available - I'll keep this issue open while I work through the issue

**Comment by jbetancur** [comment]
@mercurjl I'm refactoring this a bit for a 4.0 release and am realizing the 3.10.3 fix broke something with the select all rows state. I'm going to have to back out the change until I come to a solution.

The reason `onRowSelected` fires anytime the data changes is because it's attempting to keep the selected rows state in your component synced with regards to what rows are selected in RDT's state. 

In hindsight, `onRowSelected` is a misleading name for what should have been `onSelectedRowsChange`. Originally, it was actually named `onTableUpdated` prior to me moving to React hooks which was deprecated in place of `onRowSelected` in 3.x and I'm planning on removing in 4.x.

The main use case at the time (before I added the filtering examples) was to handle toggling a selected row state whenever the row was deleted (hence, the data was changed)

The bottom line: it's a design oversight on my part when I moved RDT to React hooks and I'll need to make some breaking changes in 4.0 to address this.

Is this more of an annoyance or is it affecting performance for your component?

**Comment by jbetancur** [comment]
@mercurjl I'm refactoring this a bit for a 4.0 release and I am realizing that onRowSelected is a misleading name for what should have been onSelectedRowsChange. Originally, it was actually named onTableUpdated prior to me moving to React hooks which was deprecated in place of onRowSelected in 3.x and I'm planning on removing in 4.x.

The main use case at the time (before I added the filtering examples) was to handle toggling a selected row state whenever the row was deleted (hence, the data was changed)

The bottom line: it's a design oversight on my part when I moved RDT to React hooks and I'll need to make some breaking changes in 4.0 to address the naming of onRowSelected as well as edges cases where it should not always fire.

Is this more of an annoyance or is it affecting performance/operation for your component?

**Comment by jbetancur** [comment]
See Release notes for 4.0.0. I've changed the way things behave selected row updates behave. The short story is that you should now use `onSelectedRowsChange` - which also will no longer fire on data changes, but as intended (when selected row state changes)

**Comment by mercurjl** [comment]
It was more of an annoyance for my project. I never had a significant amount of data in the table so I'm not entirely sure if performance would have been impacted
